src/core/allow-bind.c

```diff
@@ -0,0 +1,143 @@
+                                "CPU architecture %s is not supported",
+                                architecture_to_string(arch));
+                return 0;
+        }
+
+        r = cg_unified_controller(SYSTEMD_CGROUP_CONTROLLER);
+        if (r < 0)
+                return log_error_errno(r, "Can't determine whether the unified hierarchy is used: %m");
+
+        if (r == 0) {
+                log_debug_errno(SYNTHETIC_ERRNO(EOPNOTSUPP),
+                                "Not running with unified cgroup hierarchy, BPF is not supported");
+                return 0;
+        }
+
+        if (!bpf_probe_prog_type(BPF_PROG_TYPE_CGROUP_SOCK_ADDR, /*ifindex=*/0)) {
+                log_debug_errno(SYNTHETIC_ERRNO(EOPNOTSUPP),
+                                "BPF program type cgroup_sock_addr is not supported");
+                return 0;
+        }
+        /*
+         * Load bpf object w/o populating maps to probe whether
+         * BPF_PROG_TYPE_CGROUP_SOCK_ADDR program type is supported by kernel and
+         * if resource limits permit locking enough memory.
+         */
+        r = bpf_object_new(allow_bind_hexdump_buffer, sizeof(allow_bind_hexdump_buffer), &obj);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to create BPF object from hexdump buffer: %m");
+
+        r = bpf_object_load(obj);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to load BPF object: %m");
+
+        return 1;
+}
+
+int allow_bind_install(Unit *u) {
+        _cleanup_(bpf_object_freep) struct bpf_object *object = NULL;
+        _cleanup_(set_freep) Set *progs = NULL;
+        CGroupContext *cc;
+        int r;
+
+        cc = unit_get_cgroup_context(u);
+        if (!cc)
+                return 0;
+
+        if (cc->allow_bind_ports) {
+                r = prepare_bpf_object(u, cc->allow_bind_ports, &object);
+                if (r < 0)
+                        return log_unit_error_errno(u, r, "Failed to prepare BPF object: %m");
+
+                r = bpf_object_get_programs(object, &progs);
+                if (r < 0)
+                        return log_unit_error_errno(u, r, "Failed to get BPF programs from BPF object: %m");
+        }
+
+        r = cgroup_bpf_detach_programs(u, u->allow_bind_progs);
+        if (r < 0)
+                return log_unit_error_errno(u, r, "Failed to detach allow_bind BPF programs: %m");
+
+        /* Swap pointers to close FDs of detached programs on exit. */
+        SWAP_TWO(u->allow_bind_progs, progs);
+
+        /* AllowBind is the only feature using cgroup/bind{4|6} hooks, attaching with NONE attach flag. */
+        r = cgroup_bpf_attach_programs(u, progs, /*attach_flags*/0);
+        if (r < 0)
+                return log_unit_error_errno(u, r, "Failed to attach allow_bind BPF programs: %m");
+
+        /* If attach succeeds, store FDs of loaded programs in unit. */
+        SWAP_TWO(progs, u->allow_bind_progs);
```

Comment:
I think this logic is broken and the right eBPF programs are never attached. I tried using `bpftool prog list` and I don't see any `cgroup_sock_addr` program. 

I think it's broken because: 

- `SWAP_TWO` is executed twice, it ends up releasing the new programs. 
- `cgroup_bpf_attach_programs` is executed after `SWAP_TWO(u->allow_bind_progs, progs)`, old programs are attached again. (if they are `null` then nothing happens because there is a `SET_FOREACH` in `cgroup_bpf_attach_programs`)

I tried by adding a second aux variable and it works for me. Could you please check if a similar logic makes sense to you?

```
$ git diff src/core/allow-bind.c
diff --git src/core/allow-bind.c src/core/allow-bind.c
index 3b866bf9d9..d6103de3a9 100644
--- src/core/allow-bind.c
+++ src/core/allow-bind.c
@@ -96,7 +96,8 @@ int allow_bind_supported(void) {
 
 int allow_bind_install(Unit *u) {
         _cleanup_(bpf_object_freep) struct bpf_object *object = NULL;
-        _cleanup_(set_freep) Set *progs = NULL;
+        _cleanup_(set_freep) Set *new_progs = NULL;
+        _cleanup_(set_freep) Set *old_progs = NULL;
         CGroupContext *cc;
         int r;
 
@@ -109,7 +110,7 @@ int allow_bind_install(Unit *u) {
                 if (r < 0)
                         return log_unit_error_errno(u, r, "Failed to prepare BPF object: %m");
 
-                r = bpf_object_get_programs(object, &progs);
+                r = bpf_object_get_programs(object, &new_progs);
                 if (r < 0)
                         return log_unit_error_errno(u, r, "Failed to get BPF programs from BPF object: %m");
         }
@@ -119,15 +120,15 @@ int allow_bind_install(Unit *u) {
                 return log_unit_error_errno(u, r, "Failed to detach allow_bind BPF programs: %m");
 
         /* Swap pointers to close FDs of detached programs on exit. */
-        SWAP_TWO(u->allow_bind_progs, progs);
+        SWAP_TWO(u->allow_bind_progs, old_progs);
 
         /* AllowBind is the only feature using cgroup/bind{4|6} hooks, attaching with NONE attach flag. */
-        r = cgroup_bpf_attach_programs(u, progs, /*attach_flags*/0);
+        r = cgroup_bpf_attach_programs(u, new_progs, /*attach_flags*/0);
         if (r < 0)
                 return log_unit_error_errno(u, r, "Failed to attach allow_bind BPF programs: %m");
 
         /* If attach succeeds, store FDs of loaded programs in unit. */
-        SWAP_TWO(progs, u->allow_bind_progs);
+        SWAP_TWO(new_progs, u->allow_bind_progs);
 
         return 0;
 }
```